### DISCUSSION TITLE
[Feature suggestion] Namespaces via modules

### DISCUSSION THREAD

**Discussion by mgkuhn** [fonsp/Pluto.jl#2713] [Pluto development]
When we ask students to submit homework consisting of several exercises in form of a single Pluto notebook (exported to PDF), they often end up submitting something like:
```
using Markdown
using InteractiveUtils

# ╔═╡ b24422c3-a97f-4489-a5a2-01ef9f3b1dae
using Plots

# ╔═╡ aedb8f12-6f3f-4d2c-8555-a232845da178
md"# Exercise 1"

# ╔═╡ 917e9428-8317-11ee-3834-c9637d284c61
x1 = 0:0.1:2π;

# ╔═╡ f45623eb-45b7-4ca7-8148-2f45ffb99a8e
y1 = sin.(x1);

# ╔═╡ 76599335-5f8a-4c09-a436-a855f922aab1
plot(x1,y1)

# ╔═╡ 023ac283-e597-43ff-8a6b-40115ef340b9
md"# Exercise 2"

# ╔═╡ 7ad2bab5-57be-4393-8aa4-d977c8d4209f
x2 = 0:0.1:4π;

# ╔═╡ c35df1d1-c333-4ebf-a273-9670cb8dfc56
y2 = sin.(x2);

# ╔═╡ ebfae177-84a3-41be-b551-9bc6b1b55349
plot(x2,y2)
```

Note how they used Markdown headings to split a single notebook into separate sections, one for each exercise, but then had to (slightly awkwardly) rename common variable names, such as `x` and `y`, into `x1`, `y1`, `x2`, `y2`, as both exercises share the same global namespace.

Now there is a workaround for that, namely `let ... end` blocks, to place `x` and `y` into a local scope. However that then hinders reactive debugging and experimentation, because now all related assignments end up in a single cell, and one no longer can make intermediate values visible by simply removing the trailing semicolon.

Julia has already a mechanism for switching between several namespaces for global variables: modules. I wonder if Pluto could use modules to support several namespaces within a single notebook.

At the moment, modules can't be used for that, because modules are constrained to within a single cell. So what would be needed is some sectioning or nesting mechanism that would allow modules to enclose multiple cells, to move them into their own global namespace.

This is just a vague idea right now, but one approach might be to add to Pluto a new "Section heading" cell type that contains a section title (which could also be re-used as a module name), and that at the same time starts a new module (and closes any prior module at the same nesting level). This might be encoded as something like

```
using Markdown
using InteractiveUtils

# ╔═╡ b24422c3-a97f-4489-a5a2-01ef9f3b1dae
using Plots

# ╔s╡ aedb8f12-6f3f-4d2c-8555-a232845da178
module Exercise_1
using Markdown
using InteractiveUtils
using Plots
md"# Exercise 1"

# ╔═╡ 917e9428-8317-11ee-3834-c9637d284c61
x = 0:0.1:2π;

# ╔═╡ f45623eb-45b7-4ca7-8148-2f45ffb99a8e
y = sin.(x);

# ╔═╡ 76599335-5f8a-4c09-a436-a855f922aab1
plot(x,y)

# ╔s╡ 023ac283-e597-43ff-8a6b-40115ef340b9
end
module Exercise_2
using Markdown
using InteractiveUtils
using Plots
import ..Exercise_1
md"# Exercise 2"

# ╔═╡ 7ad2bab5-57be-4393-8aa4-d977c8d4209f
x = 0:0.1:4π;

# ╔═╡ c35df1d1-c333-4ebf-a273-9670cb8dfc56
y = sin.(x);

# ╔═╡ ebfae177-84a3-41be-b551-9bc6b1b55349
plot(x,y)

# ╔s╡ fae177eb-34a8-e1b4-155b-cbc6b1b5534a
end
```

Such a section-heading cell (`╔s╡`) might do several things:
 * close any prior module (at the same level)
 * automatically convert the section title into a human-readable module name
 * repeat all the `using` expressions previously given in `Main` (such that the user doesn't have to repeat them for each section)
 * automatically `import` any previous (or even following?) modules, such that global variables from there can still be referenced
 * output the human-readable section title

The automatic import of other modules would still allow references to variables defined in other sections, e.g., if in Exercise 2, the task was to compare with the results of Exercise 1, the student could add there
```
plot!(Exercise_1.x,Exercise_1.y)
```

Only vaguely related: #1012, #2255, #1455, #1615

**Comment by aplavin** [comment]
That would be a great addition! I missed something like this quite a few times, but didn't formulate it as a consistent feature request.
